openedx/terraform-github
# Export edx GitHub organization permissions to JSON

## Description

This adds a script  that'll export user-assigned and team-assigned repository permissions from the `edx` GitHub organization. It takes as input a list of interested repositories (`edx_import/repos.txt`), queries the GitHub REST API, and generates a JSON file (`edx_import/export.json`).

Using `export.json`, we will  then generate Terraform that'll replicate the current `edx` permission assignments within the `openedx` organization, allowing the repositories to be moved with minimal disruption to edX development or to the Core Contributor Program. If possible, we want the resulting Terraform to reflect which permissions were assigned to teams versus which permissions were assigned directly to users. Even though we will not able to add most edX employees to `openedx` organization teams, we should be able to mimic the team structure via Terraform. I'm developing this in https://github.com/openedx/terraform-github/pull/2, using a custom Terraform module that I'm calling "role".

After the `edx`->`openedx` GitHub organization transition is successful, we can then refactor the permission structure, gradually scale back permission grants, and hopefully land on a more curated and thoughtful set of roles.

### Note on export.json schema

I found that GH permission levels were easier to work with if I converted them to integers. Here's the mapping back to strings:
* 0 -> No permission
* 1 -> Read access ("push")
* 2 -> Write access ("pull")
* 3 -> Maintainer access ("maintain")
* 4 -> Admin/owner access ("admin")

## Test instructions

* Check out this branch.
* Make and activate a virtualenv.
* From the repo root, install dev requirements: `make edx_import_install_dev`.
* Optional: See if you can lint this script: `make edx_import_lint`.
* Get a GitHub personal access token from Account Settings > Developer Settings. It only needs read permissions.
* Copy the token and save it to `edx_import/github_token`.
* Optional: Limit your export dataset in order to speed up the script. In `edx_import/github_to_json.py`:
  * change `DEBUG_TEAM_SLUG_PREFIX` to something that'll capture a limited subset of teams.
  * change `DEBUG_REPO_NAME_PREFIX` to something that'll capture a limited subset of repositories.
  * Example: I changed `DEBUG_REPO_NAME_PREFIX` to `"xblock-"` but kept `DEBUG_TEAM_SLUG_PREFIX` as `""`. This allowed me to run an accurate permission dump on just xblock- repos in ~120 seconds.
  * Note: limiting the Teams dataset will make the script think that any excluded team permissions are directly-assigned user permissions.
 * Run `make edx_import_do_export`. Wait.
 * Check out `edx_import/export.json` -- see if it looks sane to you. 

## Supporting information

This begins to implement ["Back up Current permissions and be able to apply them to the repo once it's in the new org"](https://github.com/openedx/decoupling/issues/91) (internal ticket). This work in this PR is built upon in https://github.com/openedx/terraform-github/pull/2

## Other information

The script is lint-able and type-checkable (using the Makefile), but I don't plan on formally building the linting into this repo's CI, since I plan on deleting the entire `edx_import/` directory once the import is complete.

I plan on re-generating `export.json` once the rest of the PR has been reviewed in order to capture any `edx` organization changes that happened in the past few days.
